### ISSUE TITLE
CTM environment collapses to exact-zero corner on near-degenerate SU states (E=0)

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#666]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

## Summary

The forward CTM environment collapses to an **exact-zero corner** within a few renormalization sweeps on the near-degenerate iPEPS states that 2-site simple update converges to. The corner is correct after the first sweep, then the projector/renormalization step drives it to exactly `0`, so every downstream RDM and energy is `0`. Any *simple-update → forward-CTM → energy* pipeline therefore silently returns `E = 0`.

This is **not** symmetry-specific and **not** 2-site-specific:
- reproduces on a plain `DenseTensor` (independent of `SymmetricTensor` / charges / flows), and
- affects **both** single-site `ctm_tensor` and `ctm_tensor_2site` (both `recipe="1x1"` and `"2x2"`).

## Mechanism (measured)

Site tensor = `heisenberg_u1sz_init_pair(D=3)` after 40 steps of 2-site simple update (SU drives it to `nnz = 3/162`, `|A|=1`). Corner `C1` norm vs. CTM sweep count:

| max_iter | single-site `ctm_tensor` | `ctm_tensor_2site` (1x1) | `ctm_tensor_2site` (2x2) |
|---|---|---|---|
| 1  | 1.0000 | 1.0000 | **0** |
| 3  | **0** | **0** | 0 |
| 5+ | 0 | 0 | 0 |

So the corner is valid after 1 sweep and collapses to exactly zero by sweep ~3 (`1x1`) or sweep 1 (`2x2`). A **full-rank random** tensor does *not* collapse (corner ≈1 for hundreds of sweeps).

**Scope note on the trigger (not fully isolated):** naive low rank is *not* sufficient — a hand-built diagonal rank-3 site tensor is stable through the CTM. The trigger is the specific near-degenerate double-layer structure the SU converges to. The likely culprit is the projector truncation / corner renormalization producing an empty (or vanishing-norm) kept subspace on a near-singular fixed point, but the exact algebraic condition still needs pinning down.

## Reproducer (uses the actual SU path — verified to collapse)

```python
import jax, jax.numpy as jnp, numpy as np
jax.config.update("jax_enable_x64", True)
from tenax import ctm_tensor
from tenax.core.tensor import DenseTensor
from tenax.algorithms.ipeps import heisenberg_gate_u1sz, heisenberg_u1sz_init_pair
from tenax.algorithms.ipeps_simple_update import (
    _make_trotter_gate_tensor,
    _simple_update_2site_horizontal_tensor as suh,
    _simple_update_2site_vertical_tensor as suv,
)

D = 3
A, B = heisenberg_u1sz_init_pair(D=D, key=jax.random.PRNGKey(0))
H = heisenberg_gate_u1sz(); gate = _make_trotter_gate_tensor(H, 0.1, site_tensor=A)
lam_h = jnp.ones(D); lam_v = jnp.ones(D)
for step in range(40):
    if step % 2 == 0: A, B, lam_h = suh(A, B, gate, lam_h, lam_v, D)
    else:             A, B, lam_v = suv(A, B, gate, lam_h, lam_v, D)

Ad = DenseTensor(A.todense(), A.indices)   # collapse is NOT symmetry-related
print("nnz =", int(np.count_nonzero(np.abs(np.array(Ad.todense())) > 1e-10)))  # 3
for mi in (1, 3, 10):
    env, _ = ctm_tensor(Ad, chi=12, max_iter=mi, conv_tol=1e-12)
    print("max_iter", mi, "C1", float(jnp.linalg.norm(env.C1.todense())))  # 1 -> ~1, 3+ -> 0.0
```

End-to-end demonstration: `examples/su_symmetric_ctm_e2e.py` PART B (`ctm_tensor_2site` → `C1 = 0`, `E/site = 0`).

## Impact

- Any gradient-free SU → forward-CTM energy workflow returns `E = 0` once SU has driven the state toward the degenerate fixed point.
- Not caught by tests: `test_2site_symmetric_converges` only checks corner-SV **convergence**, never the energy value; an all-zero corner "converges" trivially.

## Expected behavior

CTM should degrade gracefully on near-degenerate inputs — return a finite, valid environment (and physical energy) or raise — never silently produce an exactly-zero environment. Likely fix area: projector construction / corner renormalization in `_ctm_projector.py` / `_ctm_tensor_moves.py` (floor/clamp near-zero singular values rather than truncating the kept subspace to empty).

## Suggested follow-ups

- Add a regression test asserting a **nonzero, physical** energy (not just `isfinite` / convergence) after CTM on a near-degenerate state.
- Related trigger source: the companion simple-update-collapse issue (SU drives the state to `nnz = 3/162`).

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Companion / trigger source: #667 (simple update drives the state to the near-degenerate `nnz=3/162` tensor that collapses the CTM here).

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

## Exact algebraic trigger (root-caused)

Instrumented the single-site `ctm_tensor` sweep (D=3, χ=12) on the near-classical SU-40 tensor, tracking `C1` through every directional move. The collapse is **always at the top move (`_ctm_tensor_move_top`)**, and it is an **environment-degeneracy**, not a single-projector bug.

### Where it dies (C1 norm per move)

| method | sweep 0 | sweep 1 | sweep 2 |
|---|---|---|---|
| **svd** (default) | L/T/R/B all 1.0 | L=1.0, **T=0**, R=0, B=0 | 0 |
| **eigh** | all 1.0 | all 1.0 | L=1.0, **T=0**, R=0, B=0 |
| **qr** | all 1.0 | all 1.0 | L=1.0, **T=0**, R=0, B=0 |

All three die at `moveT`; SVD dies one sweep earlier than eigh/qr.

### The mechanism (SVD / default path — proven)

`move_top` builds two enlarged corners over the shared vertical boundary:
`C1g = fuse(C1·T4)` and `C2g = fuse(C2·T2)`, then the Fishman projector uses
`M = C1g† C2g = U S V†`, `P_1 = C2g·V·S^{-1/2}`, `P_2 = C1g·U·S^{-1/2}`.
Applying `P_1` back gives `C1_new = S^{1/2} U†`, so **`C1_new = 0 ⟺ M = 0`.**

Measured `‖M‖` right before `moveT`:

| | ‖C1g‖ | ‖C2g‖ | ‖M‖ | singular values of M |
|---|---|---|---|---|
| sweep 0 | 1.00 | 1.73 | 1.34 | `[1.34, 0, 0, 0, 0]` — already **rank-1** |
| sweep 1 | 1.00 | 1.20 | **0.0** | `[0, 0, 0, 0, 0]` — **exactly zero** |

So both enlarged corners stay individually well-normalized, but their **cross-overlap `M = C1g†C2g` collapses to exactly zero** — the two corners become orthogonal over the fused boundary index. Then the Fishman regularization finishes the job:

```python
_cutoff  = 1e-14 * (S_M[0] + 1e-30)   # S_M[0]=0  ->  cutoff ≈ 1e-44
_mask    = S_M > _cutoff              # all False (S_M all exactly 0)
S_rsqrt  = jnp.where(_mask, 1/sqrt(S), 0.0)   # -> all zeros
P1 = C2g @ (V * S_rsqrt); P2 = C1g @ (U * S_rsqrt)   # -> identically zero
```

Identically-zero projectors annihilate the corner (`C1_new = C2g @ 0 = 0`), and a zero corner stays zero forever. `eps_T` is reported as `0.0` throughout, so nothing signals the failure.

### Why eigh/qr also collapse (one sweep later)

The dense eigh path uses `ρ = C1g·C1g† + C2g·C2g†`, which is nonzero for nonzero corners — so it is **not** killed by the `M=0` mechanism directly, and indeed survives one extra sweep. But the underlying environment fixed point is genuinely degenerate (a near-product state), and by sweep 2 the same corner-orthogonality defeats the eigh/qr corner reconstruction at `moveT` too. This shows the defect is an **environment degeneracy** (the CTM cannot represent the environment of a near-classical/product state), not merely the SVD relative-cutoff.

### Fix directions

1. **Defensive (CTM robustness):** detect a rank-collapsing enlarged corner (e.g. `S_M[0]` below an *absolute* floor, or `‖M‖ ≈ 0`) at the projector step and either (a) fall back to an identity/keep-existing-corner move, or (b) raise a clear diagnostic — never silently emit a zero projector. An absolute floor in the Fishman cutoff (`max(1e-14·S_M[0], abs_floor)`) stops SVD dying first, but does **not** fix eigh/qr, so this alone is insufficient.
2. **Real fix:** a robust CTM should compute the environment of a (near-)product state and return its physical energy (a classical Néel product has a perfectly well-defined environment). Compare variPEPS, which handles product states. The upstream trigger is the SU collapse in the companion issue (state driven to `nnz=3/162`).

Reproducer: instrument `ctm_tensor` per-move as above, or run `examples/su_symmetric_ctm_e2e.py` PART B.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

## Tried the absolute floor in the Fishman cutoff — does NOT fix it (negative result)

Applied an absolute floor to the dense Fishman path:
```python
_FISHMAN_ABS_FLOOR = 1e-12
_cutoff = jnp.maximum(1e-14 * (S_M[0] + 1e-30), _FISHMAN_ABS_FLOOR)
```
Result on the SU-40 tensor: `C1` still collapses to `0` by sweep 3, `E` still `0`. **No effect.**

### Why it cannot work (proven, not just observed)

`‖M‖` at the fatal move is **exactly** `0.0` (measured `maxabs = 0.000e+00`, a structural zero — the two enlarged corners are exactly orthogonal over the fused bond), not a `~1e-16` underflow. The Fishman apply yields `C1_new = (S_rsqrt · S)·U† ∝ √S`. Any floor changes `S_rsqrt`, but it always multiplies the **true** `S`, so `S=0 ⇒ C1_new=0` for every floor. Minimal demonstration with deliberately orthogonal corners:

| floor | ‖C1_new‖ |
|---|---|
| 0 | 1.8e-08 |
| 1e-12 | 2.9e-10 |
| 1e-6 | 2.9e-13 |
| 1e-2 | 2.9e-15 |

The corner gets **smaller** with a larger floor — the opposite of a fix. Same conclusion for regularizing `M` itself (`M + εI`): the corner *update* contracts the real corners `C2g†C1g = M^H`, which is still zero, so `C1_new` stays zero.

### Implication

Projector-side regularization is a dead end for this failure — `C1_new` is fundamentally the overlap-weighted projection, and the overlap is exactly zero. The fix must be one of:
1. **Degenerate-corner fallback:** detect `‖M‖ ≈ 0` (rank-collapsed enlarged corner) and skip the truncating move — keep the corner (identity move) or raise a clear diagnostic — instead of silently emitting a zero projector + zero energy.
2. **Upstream (preferred):** prevent the corners from becoming exactly orthogonal. A true product state has a rank-1 but **nonzero** overlap; `M` reaching *exactly* 0 looks self-inflicted by the rank-1 projector at sweep 0 assigning inconsistent gauges/subspaces to the left vs right corners. Worth checking whether the corner/edge gauge handling drives `M` to a hard zero (a CTM consistency issue), rather than treating the product-state environment as intrinsically unrepresentable.

The experimental branch was not kept (the floor changes numerics on healthy runs without fixing the bug).

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

## Option 2 (fix the projector gauge so M stays nonzero) — prototyped, does NOT work

Two prototypes + a decisive rank measurement show the corner orthogonalization is **intrinsic to this tensor's environment**, not a projector-gauge artifact.

### First a reframing: CTM handles product states fine

A **clean product state** (`A = v_u⊗v_d⊗v_l⊗v_r⊗|+⟩`) runs to `C1≈1.0`, `E=0.5` (correct). So "CTM can't represent product-state environments" is **false**. The SU-40 tensor is *not* a plain product state: its physical leg is frozen to `|0⟩` (all 3 nonzero entries have phys=0) with **entangled virtual bonds** (u always=2; d,l,r vary in a correlated 3-term pattern). That specific virtual structure is the trigger.

### Prototype A — rank-adaptive truncation (drop null directions)

The mask that zeroes `S^{-1/2}` for null singular values leaves the two cross-projectors with zero columns, so `P_1†P_2` is only a *partial* identity — an incomplete resolution of identity on the truncated bond. Fix attempt: truncate to the true rank `k = #{S > floor}` so `P_1†P_2 = I` exactly.

Result: **breaks the fixed-χ contraction** — `dot_general requires contracting dimensions to have the same shape, got (108,) and (9,)`. The env absorption assumes a uniform χ bond; variable per-move rank mismatches the adjacent edges. To keep fixed χ you must *pad* the null space with an orthonormal completion — which is exactly what the **eigh** isometry already does.

### Prototype B / decisive measurement — eigh already spans the full ρ, and still collapses

Instrumented the eigh path (`ρ = C1g C1g† + C2g C2g†`, a proper orthonormal isometry) at each `moveT`:

| sweep | rank(C1g) | rank(C2g) | rank(ρ) | χ | ‖C1g†C2g‖ |
|---|---|---|---|---|---|
| 0 | 2 | 1 | 3 | 12 | 1.03 |
| 1 | 1 | 4 | 5 | 12 | **0.0** |
| 2 | **0** | 1 | 1 | 12 | 0.0 |

- **`rank(ρ) ≤ χ` at every step** → eigh's top-χ isometry spans the *entire* environment density matrix; it is not truncating C1g away. A correct fixed-χ isometry (the endpoint of any "consistent gauge") is already in use.
- The corners **lose rank progressively** (`rank C1g: 2 → 1 → 0`) and become orthogonal (`‖C1g†C2g‖ = 0`) **already at sweep 1**. By the time the projector runs the corner has already lost its information — no gauge choice can preserve what is no longer there.

### Verdict

Projector-side gauge fixes (rank-adaptive, orthonormal padding, or eigh isometry) **cannot** fix this: the environment tensors degenerate progressively toward null regardless of projector method, seeded by the pathological SU tensor. The two viable fixes remain:

1. **Defensive guard (this issue):** detect the progressively-degenerating / orthogonalizing corner (e.g. `rank(C1g)` dropping or `‖M‖→0`) and bail with a clear diagnostic (or hold the last valid env) instead of silently returning a zero environment and `E=0`.
2. **Upstream cure (#667):** don't produce the pathological state — the SU collapse to the phys-frozen, virtually-entangled classical config is the root trigger.

Prototype branches were not kept.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Closing as resolved, verified by direct measurement on current main (GPU + CPU, state rebuilt from the exact reproducer).

**The silent E=0 signature no longer reproduces.** The corner no longer collapses to exact zero: with renormalization it converges to a normalized corner, and the SU trigger is gone (#667/#844: simple update no longer converges to the product state).

**The default recipe handles this state.** Running the reproducer state through both recipes, chi in {8,16,24}, max_iter=80:

| recipe | collapsed? | energy |
|---|---|---|
| 1x1 | True (rank-1) | -0.1514, chi-flat |
| 2x2 (default) | **False** (full rank) | +0.4616, chi-stable |

So the rank-1 collapse is a **1x1-recipe artifact** (deprecated and closed as wontfix in #911); the default 2x2 recipe (988c2a8, #723; #726/#746) does not collapse. The earlier "collapsed=True on 2x2" observation was at max_iter=10 -- an under-converged transient on this near-degenerate state, not the fixed point.

**Any collapse is now detected, not silent.** `env_is_collapsed()` returns True and `check_ctm_env()` emits a RuntimeWarning naming the chi_eff=1 mean-field boundary and pointing at recipe=2x2 (#747, ae01a69); the convergence criterion no longer certifies a dead corner (#898, 66232b0).

Residual (not blocking): `check_ctm_env` warns rather than raising, so a caller that ignores warnings can still read an energy off a rank-1 env -- advisory, not enforced. If hard enforcement is wanted that is a separate, smaller feature.
